# Рекомендательная система

### Первоначальный анализ датасета
Проанализируем набор данных: посмотрим на количество записей, количество признаков, пустые значения, типы данных и т.д.

### Разработка API
Разработайте программный интерфейс для итоговой модели машинного обучения. API должен позволить приложению пользователя получать: 

1)	списка топ-10 популярных фильмов;
2)	доступ к рекомендательной системе для прогнозирования списка 10 рекомендованных фильмов по жанру;
3)	доступ к рекомендательной системе для прогнозирования списка 10 рекомендованных фильмов по контенту

### Разработка приложения
Разработайте приложение с графическим интерфейсом, которое должно c помощью разработанного API показывать списки рекомендаций по выбранным пунктам.
Приложение должно предоставлять справку по имеющимся командам и их параметрам

# 1. Первоначальный анализ датасета

## 1.1 Загрузка данных

In [1]:
import pandas as pd
import numpy as np

In [2]:
films_df=pd.read_csv("IMDB_movies.csv", low_memory=False)
netflix_df=pd.read_csv("netflix_data.csv", low_memory=False)

In [45]:
films_df.head(10)

Rank                    Title                       Genre  \
0     1  Guardians of the Galaxy     Action,Adventure,Sci-Fi   
1     2               Prometheus    Adventure,Mystery,Sci-Fi   
2     3                    Split             Horror,Thriller   
3     4                     Sing     Animation,Comedy,Family   
4     5            Suicide Squad    Action,Adventure,Fantasy   
5     6           The Great Wall    Action,Adventure,Fantasy   
6     7               La La Land          Comedy,Drama,Music   
7     8                 Mindhorn                      Comedy   
8     9       The Lost City of Z  Action,Adventure,Biography   
9    10               Passengers     Adventure,Drama,Romance   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   
5  European mercenaries searching for black powde...           Yimou Zhang   
6  A jazz pianist falls for an aspiring actress i...       Damien Chazelle   
7  A has-been actor best known for playing the ti...            Sean Foley   
8  A true-life drama, centering on British explor...            James Gray   
9  A spacecraft traveling to a distant colony pla...         Morten Tyldum   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   
5      Matt Damon, Tian Jing, Willem Dafoe, Andy Lau  2016                103   
6  Ryan Gosling, Emma Stone, Rosemarie DeWitt, J....  2016                128   
7  Essie Davis, Andrea Riseborough, Julian Barrat...  2016                 89   
8  Charlie Hunnam, Robert Pattinson, Sienna Mille...  2016                141   
9  Jennifer Lawrence, Chris Pratt, Michael Sheen,...  2016                116   

   Rating   Votes  Revenue (Millions)  Metascore  
0     8.1  757074              333.13       76.0  
1     7.0  485820              126.46       65.0  
2     7.3  157606              138.12       62.0  
3     7.2   60545              270.32       59.0  
4     6.2  393727              325.02       40.0  
5     6.1   56036               45.13       42.0  
6     8.3  258682              151.06       93.0  
7     6.4    2490                 NaN       71.0  
8     7.1    7188                8.01       78.0  
9     7.0  192177              100.01       41.0

In [3]:
netflix_df.head(10)

show_id     type                             title  \
0      s1    Movie              Dick Johnson Is Dead   
1      s2  TV Show                     Blood & Water   
2      s3  TV Show                         Ganglands   
3      s4  TV Show             Jailbirds New Orleans   
4      s5  TV Show                      Kota Factory   
5      s6  TV Show                     Midnight Mass   
6      s7    Movie  My Little Pony: A New Generation   
7      s8    Movie                           Sankofa   
8      s9  TV Show     The Great British Baking Show   
9     s10    Movie                      The Starling   

                        director  \
0                Kirsten Johnson   
1                            NaN   
2                Julien Leclercq   
3                            NaN   
4                            NaN   
5                  Mike Flanagan   
6  Robert Cullen, José Luis Ucha   
7                   Haile Gerima   
8                Andy Devonshire   
9                 Theodore Melfi   

                                                cast  \
0                                                NaN   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...   
3                                                NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...   
5  Kate Siegel, Zach Gilford, Hamish Linklater, H...   
6  Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7  Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
8  Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...   
9  Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   

                                             country          date_added  \
0                                      United States  September 25, 2021   
1                                       South Africa  September 24, 2021   
2                                                NaN  September 24, 2021   
3                                                NaN  September 24, 2021   
4                                              India  September 24, 2021   
5                                                NaN  September 24, 2021   
6                                                NaN  September 24, 2021   
7  United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
8                                     United Kingdom  September 24, 2021   
9                                      United States  September 24, 2021   

   release_year rating   duration  \
0          2020  PG-13     90 min   
1          2021  TV-MA  2 Seasons   
2          2021  TV-MA   1 Season   
3          2021  TV-MA   1 Season   
4          2021  TV-MA  2 Seasons   
5          2021  TV-MA   1 Season   
6          2021     PG     91 min   
7          1993  TV-MA    125 min   
8          2021  TV-14  9 Seasons   
9          2021  PG-13    104 min   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   
5                 TV Dramas, TV Horror, TV Mysteries   
6                           Children & Family Movies   
7   Dramas, Independent Movies, International Movies   
8                       British TV Shows, Reality TV   
9                                   Comedies, Dramas   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...  
5  The arrival of a charismatic young priest brin...  
6  Equestria's divided. But a bright-eyed hero be...  
7  On a photo shoot in Ghana, an American model s...  
8  A talented 

In [24]:
films_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Title               1000 non-null   object 
 2   Genre               1000 non-null   object 
 3   Description         1000 non-null   object 
 4   Director            1000 non-null   object 
 5   Actors              1000 non-null   object 
 6   Year                1000 non-null   int64  
 7   Runtime (Minutes)   1000 non-null   int64  
 8   Rating              1000 non-null   float64
 9   Votes               1000 non-null   int64  
 10  Revenue (Millions)  872 non-null    float64
 11  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB


In [4]:
netflix_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


## 1.2 Анализ датасета

**Датасет films_df имеет следующие признаки:**
- Rank (ранг)
- Title (название)
- Genre (жанр)
- Description (описание)
- Director (режиссер)
- Actors (актёры)
- Year (год)
- Runtime (продолжительность)
- Rating (рейтинг)
- Votes (голоса)
- Revenue (рецензии)
- Metascore (метасчёт)

**Общая информация по датасету films_df:**
- Всего 12 признаков (атрибутов)
- Имеет 1000 записей
- Содержит 3 типа float64, 4 типа int64, 5 типов object
- Есть пустые значения в Revenue и Metascore

**Датасет netflix_df имеет следующие признаки:**
- Show_Id (айди шоу)
- Type (тип)
- Title (название)
- Director (режиссер)
- Cast (актёры)
- Country (страна)
- date_added (дата добавления)
- release_year (дата выпуска)
- Rating (рейтинг)
- duration (продолжительность)
- listed_in (жанры)
- Description (описание)

**Общая информация по датасету netflix_df:**
- Всего 12 признаков (атрибутов)
- Имеет 8807 записей
- Содержит 1 тип int64, 11 типов object
- Есть пустые значения в director, cast, country, date_added, rating, duration

In [25]:
films_df.describe()

Rank         Year  Runtime (Minutes)       Rating         Votes  \
count  1000.000000  1000.000000        1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000         113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962          18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000          66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000         100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000         111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000         123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000         191.000000     9.000000  1.791916e+06   

       Revenue (Millions)   Metascore  
count          872.000000  936.000000  
mean            82.956376   58.985043  
std            103.253540   17.194757  
min              0.000000   11.000000  
25%             13.270000   47.000000  
50%             47.985000   59.500000  
75%            113.715000   72.000000  
max            936.630000  100.000000

In [8]:
netflix_df.describe()

release_year
count   8807.000000
mean    2014.180198
std        8.819312
min     1925.000000
25%     2013.000000
50%     2017.000000
75%     2019.000000
max     2021.000000

Явных аномалий не замечено, все значения в пределах нормы.

## 1.3 Обработка датасета

Удалим записи с пустыми значениями

In [12]:
df_new_films = films_df.copy()
df_new_netflix = netflix_df.copy()

Удалим малозначимые признаки для рекомендательной системы

In [13]:
df_new_films.drop(columns=["Rank","Runtime (Minutes)","Revenue (Millions)","Director","Actors", "Metascore"],axis=1,inplace=True)
df_new_netflix.drop(columns=["show_id","type","director","cast","country", "date_added", "duration"],axis=1,inplace=True)

In [14]:
df_new_films.head(10)

Title                       Genre  \
0  Guardians of the Galaxy     Action,Adventure,Sci-Fi   
1               Prometheus    Adventure,Mystery,Sci-Fi   
2                    Split             Horror,Thriller   
3                     Sing     Animation,Comedy,Family   
4            Suicide Squad    Action,Adventure,Fantasy   
5           The Great Wall    Action,Adventure,Fantasy   
6               La La Land          Comedy,Drama,Music   
7                 Mindhorn                      Comedy   
8       The Lost City of Z  Action,Adventure,Biography   
9               Passengers     Adventure,Drama,Romance   

                                         Description  Year  Rating   Votes  
0  A group of intergalactic criminals are forced ...  2014     8.1  757074  
1  Following clues to the origin of mankind, a te...  2012     7.0  485820  
2  Three girls are kidnapped by a man with a diag...  2016     7.3  157606  
3  In a city of humanoid animals, a hustling thea...  2016     7.2   60545  
4  A secret government agency recruits some of th...  2016     6.2  393727  
5  European mercenaries searching for black powde...  2016     6.1   56036  
6  A jazz pianist falls for an aspiring actress i...  2016     8.3  258682  
7  A has-been actor best known for playing the ti...  2016     6.4    2490  
8  A true-life drama, centering on British explor...  2016     7.1    7188  
9  A spacecraft traveling to a distant colony pla...  2016     7.0  192177

In [15]:
df_new_netflix.head(10)

title  release_year rating  \
0              Dick Johnson Is Dead          2020  PG-13   
1                     Blood & Water          2021  TV-MA   
2                         Ganglands          2021  TV-MA   
3             Jailbirds New Orleans          2021  TV-MA   
4                      Kota Factory          2021  TV-MA   
5                     Midnight Mass          2021  TV-MA   
6  My Little Pony: A New Generation          2021     PG   
7                           Sankofa          1993  TV-MA   
8     The Great British Baking Show          2021  TV-14   
9                      The Starling          2021  PG-13   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   
5                 TV Dramas, TV Horror, TV Mysteries   
6                           Children & Family Movies   
7   Dramas, Independent Movies, International Movies   
8                       British TV Shows, Reality TV   
9                                   Comedies, Dramas   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...  
5  The arrival of a charismatic young priest brin...  
6  Equestria's divided. But a bright-eyed hero be...  
7  On a photo shoot in Ghana, an American model s...  
8  A talented batch of amateur bakers face off in...  
9  A woman adjusting to life after a loss contend...

In [16]:
df_new_films.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Title        1000 non-null   object 
 1   Genre        1000 non-null   object 
 2   Description  1000 non-null   object 
 3   Year         1000 non-null   int64  
 4   Rating       1000 non-null   float64
 5   Votes        1000 non-null   int64  
dtypes: float64(1), int64(2), object(3)
memory usage: 47.0+ KB


In [17]:
df_new_netflix.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   title         8807 non-null   object
 1   release_year  8807 non-null   int64 
 2   rating        8803 non-null   object
 3   listed_in     8807 non-null   object
 4   description   8807 non-null   object
dtypes: int64(1), object(4)
memory usage: 344.2+ KB


In [20]:
df_new_netflix['rating'] = df_new_netflix['rating'].fillna("PG-13")

Теперь посмотрим на количество пустых значений, чтобы убедиться что все пустые значения были удалены

In [22]:
missing_values = df_new_films.isnull().sum()
print(missing_values)

Title          0
Genre          0
Description    0
Year           0
Rating         0
Votes          0
dtype: int64


In [23]:
missing_values = df_new_netflix.isnull().sum()
print(missing_values)

title           0
release_year    0
rating          0
listed_in       0
description     0
dtype: int64


Пустых значений в датасетах больше нет

In [25]:
duplicates = df_new_films.duplicated().sum()
print(duplicates)

0


In [26]:
duplicates = df_new_netflix.duplicated().sum()
print(duplicates)

0


Дубликатов в датасетах тоже нет

In [27]:
df_new_films["Rating"].unique()

array([8.1, 7. , 7.3, 7.2, 6.2, 6.1, 8.3, 6.4, 7.1, 7.5, 7.8, 7.9, 7.7,
       6.6, 8.2, 6.7, 8. , 6.5, 5.3, 6.8, 4.7, 5.9, 6.3, 5.6, 8.6, 7.6,
       6.9, 2.7, 3.7, 5.8, 9. , 7.4, 4.1, 8.5, 8.8, 5.4, 6. , 5.7, 8.4,
       5.2, 5.5, 4.8, 3.9, 4.9, 5.1, 5. , 4.3, 4.6, 4.2, 4. , 3.2, 4.5,
       1.9, 3.5, 4.4])

In [29]:
df_new_netflix["rating"].unique()

array(['PG-13', 'TV-MA', 'PG', 'TV-14', 'TV-PG', 'TV-Y', 'TV-Y7', 'R',
       'TV-G', 'G', 'NC-17', '74 min', '84 min', '66 min', 'NR',
       'TV-Y7-FV', 'UR'], dtype=object)

Самый важный атрибут имеет нормальные значения.

Сохраним набор данных для дальнейшего использования в рекомендательных системах

In [17]:
# Записать DataFrame в CSV файл
csv_file = 'IMDB-films_new.csv'
df_new_films.to_csv(csv_file, index=False)

csv_file1 = 'netflix_new.csv'
df_new_netflix.to_csv(csv_file1, index=False)

# 2. Разработка API

In [ ]:
from flask import Flask, jsonify, request
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import pandas as pd

movies_df = pd.read_csv('IMDB_movies.csv')

app = Flask(__name__)

tfidf_vectorizer = TfidfVectorizer(stop_words='english')
tfidf_matrix = tfidf_vectorizer.fit_transform(movies_df['Description'])

@app.route('/top_movies', methods=['GET'])
def get_top_movies():
    top_movies = movies_df.nlargest(10, 'Rating')[['Title', 'Rating']].to_dict(orient='records')
    return jsonify(top_movies)

@app.route('/recommend_by_genre', methods=['GET'])
def recommend_by_genre():
    genre = request.args.get('genre')
    genre_movies = movies_df[movies_df['Genre'].str.contains(genre, case=False, na=False)]
    top_genre_movies = genre_movies.nlargest(10, 'Rating')[['Title', 'Rating']].to_dict(orient='records')
    return jsonify(top_genre_movies)

@app.route('/recommend_by_content', methods=['GET'])
def recommend_by_content():
    title = request.args.get('title')
    movie_index = movies_df[movies_df['Title'].str.lower() == title.lower()].index
    if movie_index.empty:
        return jsonify({"error": "Movie not found"}), 404
    cosine_sim = cosine_similarity(tfidf_matrix[movie_index[0]], tfidf_matrix).flatten()
    similar_indices = cosine_sim.argsort()[-11:-1][::-1]  # Get top 10 similar movies excluding the movie itself
    recommended_movies = movies_df.iloc[similar_indices][['Title', 'Rating']].to_dict(orient='records')
    return jsonify(recommended_movies)

if __name__ == "__main__":
    app.run(host="0.0.0.0", port=5000, debug=True)


# 3. Разработка приложения

In [ ]:
import streamlit as st
import requests

API_BASE_URL = "http://172.19.0.1:5000"

st.title("Movie Recommendation App")
st.write("Это приложение позволяет запрашивать рекомендации фильмов через API.")

option = st.sidebar.selectbox(
    "Выберите действие",
    ["Справка", "Топ-10 популярных фильмов", "Рекомендации по жанру", "Рекомендации по контенту"]
)

if option == "Справка":
    st.subheader("Справка")
    st.write("""
    Это приложение поддерживает следующие команды:

    1. **Топ-10 популярных фильмов**: Возвращает список из 10 фильмов с наивысшим рейтингом.
    - Путь: `/top_movies`
    - Метод: `GET`

    2. **Рекомендации по жанру**: Позволяет получить топ-10 фильмов для указанного жанра.
    - Путь: `/recommend_by_genre`
    - Метод: `GET`
    - Параметр: `genre` (например, `genre=Action`)

    3. **Рекомендации по контенту**: Находит фильмы, похожие на указанный фильм по его описанию.
    - Путь: `/recommend_by_content`
    - Метод: `GET`
    - Параметр: `title` (например, `title=Guardians of the Galaxy`)
    """)

elif option == "Топ-10 популярных фильмов":
    st.subheader("Топ-10 популярных фильмов")
    response = requests.get(f"{API_BASE_URL}/top_movies")
    if response.status_code == 200:
        top_movies = response.json()
        for movie in top_movies:
            st.write(f"**{movie['Title']}** - Рейтинг: {movie['Rating']}")
    else:
        st.error("Ошибка при получении данных")

elif option == "Рекомендации по жанру":
    st.subheader("Рекомендации по жанру")
    genre = st.text_input("Введите жанр (например, Action):")
    if genre:
        response = requests.get(f"{API_BASE_URL}/recommend_by_genre", params={"genre": genre})
        if response.status_code == 200:
            genre_movies = response.json()
            for movie in genre_movies:
                st.write(f"**{movie['Title']}** - Рейтинг: {movie['Rating']}")
        else:
            st.error("Ошибка при получении данных")

elif option == "Рекомендации по контенту":
    st.subheader("Рекомендации по контенту")
    title = st.text_input("Введите название фильма:")
    if title:
        response = requests.get(f"{API_BASE_URL}/recommend_by_content", params={"title": title})
        if response.status_code == 200:
            content_movies = response.json()
            for movie in content_movies:
                st.write(f"**{movie['Title']}** - Рейтинг: {movie['Rating']}")
        else:
            st.error("Ошибка при получении данных")

# Run this file with Streamlit by calling:
# streamlit run api_ui.py
